include/gridtools/sid/delegate.hpp

```diff
@@ -25,36 +26,39 @@ namespace gridtools {
          * @tparam Sid a object that models `SID` concept.
          */
         template <class Sid>
-        class delegate {
+        struct delegate {
+            static_assert(is_sid<Sid>::value, GT_INTERNAL_ERROR);
             Sid m_impl;
 
-            static_assert(is_sid<Sid>::value, GT_INTERNAL_ERROR);
+            template <bool IsRef = std::is_reference<Sid>::value, std::enable_if_t<!IsRef, int> = 0>
+            delegate(Sid impl) : m_impl(std::move(impl)) {}
 
-            friend ptr_holder_type<Sid> sid_get_origin(delegate &obj) { return get_origin(obj.m_impl); }
-            friend strides_type<Sid> sid_get_strides(delegate const &obj) { return get_strides(obj.m_impl); }
-            friend lower_bounds_type<Sid> sid_get_lower_bounds(delegate const &obj) {
-                return get_lower_bounds(obj.m_impl);
-            }
-            friend upper_bounds_type<Sid> sid_get_upper_bounds(delegate const &obj) {
-                return get_upper_bounds(obj.m_impl);
-            }
-
-          protected:
-            Sid const &impl() const { return m_impl; }
-            Sid &impl() { return m_impl; }
-
-          public:
-            template <class Arg>
-            explicit delegate(std::initializer_list<Arg> lst) : m_impl(*lst.begin()) {}
-
-            template <class Arg>
-            explicit delegate(Arg &&arg) noexcept : m_impl(std::forward<Arg>(arg)) {}
+            template <bool IsRef = std::is_reference<Sid>::value, std::enable_if_t<IsRef, int> = 0>
+            delegate(Sid impl) : m_impl(impl) {}
         };
 
         template <class Sid>
-        ptr_diff_type<Sid> sid_get_ptr_diff(delegate<Sid> const &);
+        decltype(sid_get_origin(std::declval<Sid &>())) sid_get_origin(delegate<Sid> &obj) {
```

Comment:
It's easy to overlook that this works as a SFINAE-enabler. Maybe put a comment.

Reply:
done